src/cpp/src/sampler.cpp

```diff
@@ -587,45 +590,95 @@ SamplerOutput Sampler::sample(std::vector<SequenceGroup::Ptr> & sequence_groups,
 
         const void * sequence_group_logits_data = logits_data + vocab_size * currently_processed_tokens;
         ov::Tensor sequence_group_logits(ov::element::f32, ov::Shape{num_running_sequences, actual_seq_len, vocab_size}, (void *)sequence_group_logits_data);
-
+        size_t decrease_len = 0;
         if (sequence_group->requires_sampling()) {
+            auto token_id = actual_seq_len;
+            // prompt phase
+            if (sequence_group->get_num_processed_tokens() == 0) {
+                token_id -= (sequence_group->get_prompt_len() - 1);
+            }
             if (sampling_params.is_greedy_decoding() || sampling_params.is_multinomial()) {
                 std::vector<Sequence::Ptr> running_sequences = sequence_group->get_running_sequences();
-                if (sampling_params.is_greedy_decoding()) {
-                    OPENVINO_ASSERT(num_running_sequences == 1);
+                // limitation for speculative decoding
+                if (is_validation_mode_enabled || sampling_params.is_greedy_decoding()) {
+                    OPENVINO_ASSERT(running_sequences.size() == 1);
```

Comment:
speculative decoding should work with parallel sampling as well, right? in a theory

Reply:
This limitation was introduced because we were not sure how to handle speculative decoding with several sequences in case of different match number. I have remove this assert. But my suggestion is to check this case in real scenario and return this condition in case of unexpected behavior with TODO and link to the ticket